app/components/search_bar/search_bar.js

```diff
@@ -0,0 +1,390 @@
+        this.searchContainerRef = ref;
+    }
+
+    setInputKeywordRef = (ref) => {
+        this.inputKeywordRef = ref;
+    }
+
+    blur = () => {
+        this.inputKeywordRef.blur();
+    };
+
+    focus = () => {
+        this.inputKeywordRef.focus();
+    };
+
+    onBlur = async () => {
+        if (this.props.leftComponent) {
+            await this.collapseAnimation();
+        }
+        this.props.onBlur();
+    };
+
+    onLeftComponentLayout = (event) => {
+        const leftComponentWidth = event.nativeEvent.layout.width;
+        this.setState({leftComponentWidth});
+    };
+
+    onSearch = async () => {
+        if (this.props.keyboardShouldPersist === false) {
+            await Keyboard.dismiss();
+        }
+
+        this.props.onSearchButtonPress(this.props.value);
+    };
+
+    onChangeText = (text) => {
+        if (this.props.onChangeText) {
+            this.props.onChangeText(text);
+        }
+    };
+
+    onFocus = () => {
+        InteractionManager.runAfterInteractions(async () => {
+            if (this.props.leftComponent) {
+                await this.expandAnimation();
+            }
+
+            if (this.props.onFocus) {
+                this.props.onFocus();
+            }
+        });
+    };
+
+    onClear = () => {
+        this.focus();
+        this.props.onChangeText('', true);
+    };
+
+    onCancel = () => {
+        Keyboard.dismiss();
+        InteractionManager.runAfterInteractions(() => {
+            if (this.props.onCancelButtonPress) {
+                this.props.onCancelButtonPress();
+            }
+        });
+    };
+
+    onSelectionChange = (event) => {
+        this.props.onSelectionChange(event);
+    };
+
+    expandAnimation = () => {
+        return new Promise((resolve) => {
+            Animated.parallel([
+                Animated.timing(
+                    this.leftComponentAnimated,
+                    {
+                        toValue: 100,
+                        duration: 200,
+                    },
+                ),
+                Animated.timing(
+                    this.searchContainerAnimated,
+                    {
+                        toValue: this.state.leftComponentWidth * -1,
+                        duration: 200,
+                    },
+                ),
+            ]).start();
+            resolve();
+        });
+    }
+
+    collapseAnimation = () => {
+        return new Promise((resolve) => {
+            Animated.parallel([
+                Animated.timing(
+                    this.leftComponentAnimated,
+                    {
+                        toValue: 0,
+                        duration: 200,
+                    },
+                ),
+                Animated.timing(
+                    this.searchContainerAnimated,
+                    {
+                        toValue: 0,
+                        duration: 200,
+                    },
+                ),
+            ]).start();
+            resolve();
```

Comment:
do we want to resolve when the animation starts or when the animation ends? if the latter this needs to be called in the `start` callback.